IGroupForm improvements

davidread opened the pull request:
Derived from https://github.com/ckan/ckan/issues/2344

davidread commented on ckan/controllers/group.py:
```diff
@@ -88,8 +85,6 @@ def _bulk_process_template(self, group_type):
     def _replace_group_org(self, string):
         ''' substitute organization for group if this is an org'''
-        if self.group_type == 'organization':
-            string = re.sub('^group', 'organization', string)
```
Only the OrganizationController has customized logic functions - it is not to do with the group_type. Therefore this code has been moved into OrganizationController.

davidread commented on ckan/controllers/group.py:
```diff
@@ -104,22 +99,25 @@ def _render_template(self, template_name):
+    def _redirect_to__this_controller(self, *args, **kw):
+        ''' wrapper around redirect_to but it adds in this request's controller
+        (so that it works for Organization or other derived controllers)'''
+        kw['controller'] = request.environ['pylons.routes_dict']['controller']
```
redirects and url_for should keep the same controller as the current request. That is better than trying to determine it from the group_type, which may be a custom one.

wardi replied:
I've never seen a double underscore in the middle of a method name. Is there some significance I'm missing?

davidread replied:
Mild significance... just a separator, like two spaces. This method is a variation on the familiar func 'redirect_to'. Rather than reading 'redirect to this controller', it more means 'redirect to whatever the *args/**kw say, as long as it is in this controller'. I'll remove the two if you prefer or can think of a better name.

davidread commented on ckan/controllers/group.py:
```diff
@@ -133,6 +131,21 @@ def _guess_group_type(self, expecting_name=False):
+            abort(404, _('Group not found'))
+        if group.type not in self.group_types:
+            abort(404, _('Incorrect group type'))
+        return group.type
```
This method is called at the start of most controller methods, to stop it working if the user tries using /organization/mygroup instead of /group/mygroup etc.

davidread commented on ckan/controllers/group.py:
```diff
@@ -295,15 +305,17 @@ def pager_url(q=None, page=None):
-            for plugin in plugins.PluginImplementations(plugins.IFacets):
-                if self.group_type == 'organization':
+            from ckan.controllers.organization import OrganizationController
+            if isinstance(self, OrganizationController):
```
The choice between group_facets and organization_facets is down to whether we're in GroupController or OrganizationController. Ideally this would be a method which OrganizationController overrides, but it is neater to keep it inline. So we simply use isinstance.

wardi replied:
then we have three ways testing if this is related to an organization. that doesn't seem good.

davidread replied:
On the face of it these three are the same. I think I've been having half an eye on the situation where someone wants to inherit and customize OrganizationController. But really, rather than hypothecate wildly, I think we should just leave these alone until we have a proper use case and test, and then improve these. Otherwise my head will explode and will get it wrong anyway. What do you think?

wardi replied:
No, I'm confusing cases where we're looking at a specific org/group vs this controller search method. I agree with your comment that this would be nicer as a method overridden in OrganizationController, though. Hiding an isinstance checks in here makes it hard to discover the special case.

davidread commented on ckan/controllers/group.py:
```diff
@@ -395,7 +405,7 @@ def bulk_process(self, id):
             # unicode format (decoded from utf8)
             limit = 500
-            self._read(id, limit)
+            self._read(id, limit, group_type)
```
Pass in group_type, instead of working it out again in _read.

davidread commented:
@wardi Do you have a chance to cast an eye over this and see if it is generally ok? I'm not sure if this completes IGroupForm, but it certainly now works in the basic cases of editing and viewing them.

Since people will use the default GroupController, I guess it is worth changing the group templates so that all the links are to /<group_type>/new rather than /group/new? Otherwise they'd have to provide custom templates for everything.

wardi commented:
@davidread yes, I'll take a look.

@deniszgonjanin have you started on some core changes to ckan for making scheming work with groups and orgs? This is surely related.

deniszgonjanin commented:
@wardi there was really only one change needed to make scheming work, and that was for IGroupForm to pass the group_type to the plugin. However that change was backwards incompatible the way I did it for my immediate purposes.

davidread commented:
@deniszgonjanin Could you be more specific about which function to/from
would helpfully pass group_type?


On 18 March 2015 at 12:55, Denis Zgonjanin <notifications@github.com> wrote:

> @wardi <https://github.com/wardi> there was really only one change needed
> to make scheming work, and that was for IGroupForm to pass the group_type
> to the plugin. However that change was backwards incompatible the way I did
> it for my immediate purposes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ckan/ckan/pull/2354#issuecomment-82956786>.
>

deniszgonjanin commented:
@davidread I am fairly sure it was https://github.com/ckan/ckan/blob/master/ckan/controllers/group.py#L59

`group_type` needs to be passed along to implementing plugins, for scheming to know which schema it is supposed to use.

davidread commented:
I see, thanks.

When you edit an existing group then the type is available as data_dict['type']. However if it is a new group then data_dict is {}. How about we add in the type into data_dict at this point? That would save changing the variables passed to setup_template_variables, and makes it consistent between new & edit forms. It just means the setup_template_variables method can't tell if it is a new or edit by seeing if data_dict == {}.

The alternative is to try calling setup_template_variables(context, data_dict, group_type) and if that raises an exception, try it without the group_type param. That keeps it backward compatible, but is perhaps not as neat. Thoughts?

wardi commented:
As much as possible I hope these changes will bring IGroupForm use more in-line with the recent IDatasetForm changes. In particular I resisted using setup_template_variables because it's really tricky to make sure that function gets called (and nothing else messes with a value in the context) before you need to read the value out for all cases.

Instead I explicitly pass dataset_type to the jinja2 render calls. It's a larger change but it feels less magical and easier to understand. https://github.com/ckan/ckan/pull/1795

davidread commented:
Thanks @wardi - I've now added group_type to all the render calls.
https://github.com/ckan/ckan/commit/70e51adb8195d27c1f9714ca70c71133e8fc906b

davidread commented:
It would be great to merge this at some point if it is ok @wardi

wardi commented:
@davidread yes, thanks for the reminder! I'll look at it now

wardi commented on ckan/controllers/group.py:
```diff
@@ -227,15 +240,15 @@ def _read(self, id, limit):
         def search_url(params):
             if group_type == 'organization':
```
@davidread shouldn't this be checking is_organization instead? I ask because I see you fixed some code near this

davidread replied:
I think group['is_organization'] and group_type=='organization' are interchangeable. I don't see a situation where you'd have one and not the other - or can you?

wardi replied:
IGroupForm.group_types lets plugins define different types for their orgs. I take that to mean is_organization is the definitive test. I haven't actually tried creating different org types, but that's what I hope to do with scheming in the future: make orgs as customizable as datasets, and one day maybe users too.

wardi commented on ckan/controllers/group.py:
```diff
@@ -606,19 +635,19 @@ def delete(self, id):
-                    h.flash_notice(_('Group has been deleted.'))
-                self._redirect_to(controller='group', action='index')
+                h.flash_notice(_('%s has been deleted.')
+                               % group_type.capitalize())
```
should the parameter be in its own gettext call? if so that's a hidden string for the translators to find. if not the type is not translatable.

Or how about attaching a `deleted_flash_notice_message` object to the class and let the OrganizationController override it?

davidread replied:
> should the parameter be in its own gettext call? if so that's a hidden string for the translators to find. if not the type is not translatable.

Hmm I have an idea.

> Or how about attaching a deleted_flash_notice_message object to the class and let the OrganizationController override it?

If you attach the message to the class, then you can't customize it for a different group_type unless you create a subclass GroupController, which would not normally be necessary for an IGroupForm - as you can see from ckanext/example_igroupform/plugin.py.

wardi commented on ckan/controllers/organization.py:
```diff
@@ -1,19 +1,23 @@
-    the group_type to organization so that the group controller knows that
-    it is in fact the organization controller.  All the main logical
-    differences are therefore in the group controller.
+    ''' The organization controller is for Groups of type 'organization'. It
```
`''' The organization controller is for Organizations, which are implemented as ckan groups with is_organization=True`

wardi commented on ckan/controllers/organization.py:
```diff
@@ -1,19 +1,23 @@
-    The main differences the group controller provides for organizations are
-    a few wrapper functions that swap organization for group when rendering
-    templates, redirecting or calling logic actions '''
+    Nearly all the code for both is in the GroupController (for simplicity?).
```
for historical reasons, I hope

wardi commented:
Other than the comments I added this looks great @davidread 

I'm just curious what custom org types you have added that use this.

davidread commented on ckan/controllers/group.py:
```diff
@@ -608,17 +637,22 @@ def delete(self, id):
-                self._redirect_to(controller='group', action='index')
+                else:
+                    h.flash_notice(_('%s has been deleted.')
+                                   % _(group_type.capitalize()))
```
I've put back the original flash messages, to avoid disrupting the default group_types. And although it is a hidden string here for a custom group_type, it is translatable if people spot it. Am not sure what else we can do.

davidread commented:
I've also got rid of the double underscores, in case that confuses any more.

davidread commented:
Ok that's built again now. What do you think?

davidread commented:
Ok, I've addressed the points now I think. I've not tried it with any custom org types. Merely following up on the issue raised here: https://github.com/ckan/ckan/issues/2344 and added a v. basic test for custom group type. It's all a bit theoretical use, but at least it is functional now...

davidread commented:
Please do review @wardi when you have a chance

wardi commented:
@davidread thanks for the ping, will do

wardi commented:
@davidread looks great! I'm looking forward to using this. Please resolve the conflict and I will merge.

davidread commented:
Cheers @wardi
